File: src/main/java/me/zeroeightsix/kami/setting/impl/EnumSetting.java
package me.zeroeightsix.kami.setting.impl;

import com.google.common.base.Converter;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.converter.EnumConverter;

import java.util.function.BiConsumer;
import java.util.function.Predicate;

/**
 * Created by 086 on 14/10/2018.
 */
public class EnumSetting<T extends Enum> extends Setting<T> {

    private EnumConverter converter;
    public final Class<? extends Enum> clazz;

    public EnumSetting(T value, Predicate<T> restriction, BiConsumer<T, T> consumer, String name, Predicate<T> visibilityPredicate, Class<? extends Enum> clazz) {
        super(value, restriction, consumer, name, visibilityPredicate);
        this.converter = new EnumConverter(clazz, value);
        this.clazz = clazz;
    }

    @Override
    public Converter converter() {
        return converter;
    }

}

File: src/main/kotlin/org/kamiblue/client/setting/settings/impl/primitive/EnumSetting.kt
package org.kamiblue.client.setting.settings.impl.primitive

import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import org.kamiblue.client.setting.settings.MutableSetting
import org.kamiblue.commons.extension.next
import java.util.*

class EnumSetting<T : Enum<T>>(
    name: String,
    value: T,
    visibility: () -> Boolean = { true },
    consumer: (prev: T, input: T) -> T = { _, input -> input },
    description: String = ""
) : MutableSetting<T>(name, value, visibility, consumer, description) {

    val enumClass: Class<T> = value.declaringClass
    val enumValues: Array<out T> = enumClass.enumConstants

    fun nextValue() {
        value = value.next()
    }

    override fun setValue(valueIn: String) {
        super.setValue(valueIn.toUpperCase(Locale.ROOT).replace(' ', '_'))
    }

    override fun write(): JsonElement = JsonPrimitive(value.name)

    override fun read(jsonElement: JsonElement?) {
        jsonElement?.asJsonPrimitive?.asString?.let { element ->
            enumValues.firstOrNull { it.name.equals(element, true) }?.let {
                value = it
            }
        }
    }

}